tools/c7n_azure/c7n_azure/session.py

```diff
@@ -53,19 +55,24 @@
 class Session(object):
 
     def __init__(self, subscription_id=None, authorization_file=None,
-                 resource=constants.RESOURCE_ACTIVE_DIRECTORY):
+                 cloud_endpoints=AZURE_PUBLIC_CLOUD, auth_endpoint=DEFAULT_AUTH_ENDPOINT):
         """
         :param subscription_id: If provided overrides environment variables.
         :param authorization_file: Path to file populated from 'get_functions_auth_string'
-        :param resource: Resource endpoint for OAuth token.
+        :param cloud_endpoints: List of endpoints for specified Azure Cloud. Defaults to public.
+        :param auth_endpoint: Resource endpoint for OAuth token.
         """
 
         self._provider_cache = {}
         self.subscription_id_override = subscription_id
         self.credentials = None
         self.subscription_id = None
         self.tenant_id = None
-        self.resource_namespace = resource
+        self.endpoints = cloud_endpoints or AZURE_PUBLIC_CLOUD
+        self.keyvault_auth_override = None
+        self.resource_namespace = None
+        self.resolve_auth_endpoint(auth_endpoint)
+        self.storage_endpoint = self.endpoints.suffixes.storage_endpoint
```

Comment:
I find this a bit hard to follow, a combination of the way it is resolved/used, the overrides, and the inconsistent naming.  Calling into the resolve method to set the class level attributes isn't ideal as its harder to read and test.  One proposal that I think simplifies everything a bunch:

Get rid of `keyvault_auth_override`, and just keep both the keyword and the resolved value in two fields here (rename constructor `auth_endpoint` to `resource_endpoint_type`).  Return the single value from `resolve_auth_endpoint` (and optionally rename to `get_auth_endpoint`.

```
self.cloud_endpoints = cloud_endpoints
self.resource_endpoint_type = resource_endpoint_type
self.resource_endpoint = self.get_auth_endpoint(resource_endpoint_type)
self.storage_endpoint = self.cloud_endpoints.suffixes.storage_endpoint
```

Then for the keyvault check you can just keep doing basically the old way that is cloud agnostic.
```
if self.resource_endpoint_type == constants.keyvault_something:
```